src/slurm_plugin/slurm_resources/slurm_resources.py

```diff
@@ -0,0 +1,413 @@
+        """Initialize slurm node with attributes."""
+
+    def is_node_state_healthy(self):
+        """Check if a slurm node's scheduler state is considered healthy."""
+        # Check to see if node is in DRAINED, ignoring any node currently being replaced
+        if self.is_drained():
+            logger.warning("Node state check: node %s in DRAINED, replacing node", self)
+            return False
+        # Check to see if node is in DOWN, ignoring any node currently being replaced
+        elif self.is_down():
+            if not self.is_nodeaddr_set():
+                # Silently handle failed to launch dynamic node to clean up normal logging
+                logger.debug("Node state check: node %s in DOWN, replacing node", self)
+            else:
+                logger.warning("Node state check: node %s in DOWN, replacing node", self)
+            return False
+        return True
+
+    def is_healthy(self, terminate_drain_nodes):
+        """Check if a slurm node is considered healthy."""
+        return (
+            self.is_powering_down()
+            or self.is_backing_instance_valid()
+            and (self.is_node_state_healthy() if terminate_drain_nodes else True)
+        )
+
+    def is_bootstrap_failure(self):
+        """Check if a slurm node has boostrap failure."""
+        # no backing instance + [working state]# in node state
+        if self.is_configuring_job() and not self.is_backing_instance_valid():
+            logger.warning("Node bootstrap error: Node %s is in power up state without valid backing instance", self)
+            return True
+        # Dynamic node in DOWN*+CLOUD+POWER state
+        elif self.is_resume_failed():
+            logger.warning("Node bootstrap error: Resume timeout expires for node %s", self)
+            return True
+        return False
+
+    def is_powering_down_node_with_nodeaddr(self):
+        """
+        Check if a slurm node is a powering down node with instance backing.
+
+        Nodes in powering_down(i.e. down%) state will still have the nodeaddr of the backing instance.
+        Nodeaddr is reset to nodename automatically by slurm after the power_down process,
+        when node goes back into power_saving(i.e. idle~).
+        If instance is not terminated during powering_down for some reason,
+        it becomes orphaned once the nodeaddr is reset, and will be handled as an orphaned node.
+        """
+        return self.is_nodeaddr_set() and (self.is_power() or self.is_powering_down())
+
+
+class EC2InstanceHealthState:
+    def __init__(self, id, state, instance_status, system_status, scheduled_events):
+        """Initialize slurm node with attributes."""
+        self.id = id
+        self.state = state
+        self.instance_status = instance_status
+        self.system_status = system_status
+        self.scheduled_events = scheduled_events
+
+    def _fail_ec2_health_check(self, current_time, health_check_timeout):
```

Comment:
Minor: Alternatively we can initialize the object with current_time and health_check_timeout